doc/dev/ragged_iter_domain_design_doc.md

```diff
@@ -0,0 +1,533 @@
+- **Offset**: Starting position of each component in contiguous storage
+- **Extent**: Sum of all nested domain extents (total extent)
+
+---
+
+## 4. Goals and Non-Goals
+
+### Goals
+
+- Support ragged dimensions in fusion IR for expert parallelism
+
+### Non-Goals
+
+- Multi-level nesting (ragged within ragged)
+- Multiple ragged dimensions per tensor (deferred to future work)
+- Inlining
+- Extension of single-GPU schedulers for ragged dimensions
+
+Inlining would be eventually necessary for lowering to CUDA code but should not be necessary for distributed device (multi-GPU) parallelization. Similarly, the schedulers will be left as is for now.
+
+---
+
+## 5. Requirements
+
+### Functional Requirements
+
+- **IR Representation**: Represent ragged dimensions with variable extents per batch component
+- **Offset Computation**: Compute and provide access to offsets for indexing into contiguous storage
+- **Operations Support**: Support creation and basic transformations on ragged dimensions (flatten operation TBD based on expert parallelism requirements)
+- **Code Generation**: Generate correct CUDA code with offset-based indexing for ragged iteration
+- **Uniform Properties**: Enforce uniform execution properties (ParallelType, IterType) across components
+- **Integration**: Work within existing TensorView/TensorDomain infrastructure
+
+### Non-Functional Requirements
+
+- **Backward Compatibility**: No breaking changes to existing IterDomain API or behavior
+- **Maintainability**: Clear separation of ragged-specific logic from regular IterDomain code
+
+---
+
+## 6. Proposed Design
+
+### 6.1 High-Level Architecture
+
+`RaggedIterDomain` is a direct subclass of `IterDomain`, allowing it to be used anywhere an IterDomain is expected (in TensorDomain, transformations, etc.) while providing ragged-specific functionality.
+
+```
+Val
+ └── IterDomain
+      ├── (regular IterDomain instances)
+      └── RaggedIterDomain (new subclass)
+```
+
+A `TensorDomain` can contain a mix of regular `IterDomain` and `RaggedIterDomain` instances, representing tensors with both uniform and ragged dimensions. For the initial implementation, **only one RaggedIterDomain per TensorDomain is supported** (matching PyTorch's restriction). Example:
+
+```cpp
+// Example: Batch of sequences with variable lengths
+auto batch = IrBuilder::create<IterDomain>(0, 3);
+auto ragged_seq = IrBuilder::create<RaggedIterDomain>({seq0, seq1, seq2});
+auto feature = IrBuilder::create<IterDomain>(0, 4);
+
+// TensorDomain: {batch, ragged_seq, feature}
+// Only one ragged dimension allowed
+auto tensor_domain = IrBuilder::create<TensorDomain>({batch, ragged_seq, feature});
+```
+
+The compilation pipeline detects ragged dimensions and routes to appropriate indexing and code generation paths.
+
+### 6.2 Core Abstractions
+
+#### RaggedIterDomain Class
+
+```cpp
+class RaggedIterDomain : public IterDomain {
```

Comment:
Note to myself: there's no way to find the offsets tensor from a RaggedIterDomain. Could that be a problem?